lib/smack/DSAWrapper.cpp

```diff
@@ -215,23 +118,115 @@ unsigned DSAWrapper::getPointedTypeSize(const Value *v) {
 }
 
 int DSAWrapper::getOffset(const Value *v) {
-  return getOffset(new MemoryLocation(v));
+  if (DG->hasCell(*v)) {
+    auto cell = DG->getCell(*v);
+    auto node = cell.getNode();
+    // Be consistent with the old implementation.
+    if (node->isOffsetCollapsed())
+      return -1;
+    auto offset = cell.getOffset();
+    assert(offset <= INT_MAX && "Cannot handle large offsets");
+    return offset;
+  }
+  // Return 0 if we can't find the cell that `v` associates with.
+  return 0;
+}
+
+const sea_dsa::Node *DSAWrapper::getNode(const Value *v) {
+  // For sea-dsa, a node is obtained by getting the cell first.
+  if (DG->hasCell(*v)) {
+    auto node = DG->getCell(*v).getNode();
+    assert(node && "DSNode should not be NULL.");
+    return node;
+  }
+  llvm_unreachable("Values should have cells.");
 }
 
-// TODO: Should this return the node or its leader?
-const DSNode *DSAWrapper::getNode(const Value *v) {
-  const llvm::EquivalenceClasses<const llvm::DSNode *> &eqs =
-      nodeEqs->getEquivalenceClasses();
-  auto *N = nodeEqs->getMemberForValue(v);
-  return N ? eqs.getLeaderValue(N) : nullptr;
+bool DSAWrapper::isTypeSafe(const Value *v) {
+  typedef std::unordered_map<unsigned, bool> FieldMap;
+  typedef std::unordered_map<const sea_dsa::Node *, FieldMap> NodeMap;
+  static NodeMap nodeMap;
+
+  auto node = getNode(v);
```

Comment:
Add `assert(node)` after this line as a sanity check.

Reply:
I don't think we need a sanity check here since `getNode` already gets one upon returning.